### PR TITLE
Add environment variable handling, fix collision woes, do some housekeeping

### LINKED ISSUE #31: make an easy way to toggle audio
don't wanna hear it constantly while debugging

### PR DESCRIPTION
**Pull request by rhroberts**
This PR makes the game aware of a few OS environment variables: `FMA_PROD`,`FMA_SCENE`, and `FMA_MUSIC`. 

TL;DR: I wanted to be able to start the game from a particular scene, or with the music switched off, without having to make 
temporary changes and fiddle with the code logic. Setting `FMA_PROD="true"` will make the game run "normally" and ignore all the other variables you might have set. 

For linux, an example would be running love like this:

```bash
FMA_MUSIC=false love .
```

In windows (powershell), you could do:

```powershell
$env:FMA_MUSIC="false"
C:\path\to\love.exe .
```
I'm sure this could be used in the `tasks.json` file that @halfhorst set up as well, but I haven't looked into it.

Here's the spec from `env.lua`:

```lua
local envVars = {
    FMA_PROD = {
        name = "FMA_PROD",
        type = "boolean",
        options = set:new{true, false},
        default = false,
        description = "If `true`, ignore all other environment variables. This is 'normal' gameplay."
    },
    FMA_MUSIC = {
        name = "FMA_MUSIC",
        type = "boolean",
        options = set:new{true, false},
        default = true,
        description = "Whether or not to play game music."
    },
    FMA_SCENE = {
        name = "FMA_SCENE",
        type = "string",
        options = set:new{
            "TitleScene", "EnterPatientsScene", "NavigationScene", "AlchemyScene"
        },
        default = "TitleScene",
        description = "Which scene to start the game on."
    }
}
```

Other things in this PR:

- I added a utils folder with a few odds and ends
  - implementation of sets in lua
  - privacy module with function that makes a table "read-only"
  - the env module that implements what I described above
- audio folder -- currently just has a little wrapper around `love.audio.newSource` that does the logic for checking for the FMA_MUSIC env var, to keep that out of the main scene logic
- Other stuff I can't remember! I broke the frog somehow! It randomly gets deleted and throws an error every now and then

Hit me with your thoughts

### CODE REVIEW

**Comment by rhroberts** [comment]
this will close #31

**Comment by rhroberts** [comment]
c8fa2a6 fixes the frog thing. Also made a class that both the frog and cat are made from in `borked_patient.lua`.

Still some issues with the physicker to deal with:

- [x] Collision box seems off again
- [x] Looks blurry
- [x] Can see bits of other frames in the animation

**Comment by rhroberts** [comment]
Also, jumping right to the nav scenes messes up logic with patients being there. Need to think about how best to deal with that.

**Review by dmarc3 — APPROVED** [review]
Gave all the updates a read through and they look fine to me. I haven't had a chance to pull it locally and run it but the changes you made make sense to me. I am wondering if the physicker edge stuff is because our Aesprite sprites have literally 0 buffer on the edges... so it could be Peachy cutting the images not 100% exactly right or something. Might be worth a test.

**Comment by rhroberts** [comment]
> Gave all the updates a read through and they look fine to me. I haven't had a chance to pull it locally and run it but the changes you made make sense to me. I am wondering if the physicker edge stuff is because our Aesprite sprites have literally 0 buffer on the edges... so it could be Peachy cutting the images not 100% exactly right or something. Might be worth a test.

Yeah I'm confused on this one... It looks fine on main but I checked the diff on my physicker sprite and it's the same one. Also it seems to be drawn blurry. Like everything did before we set `love.graphics.setDefaultFilter("nearest", "nearest")` in main.lua. Idk.

Anyway, I won't merge until the physicker stuff is sorted.

**Comment by dmarc3** [comment]
We might want to consider a generic "debug" mode which will draw all existing bodies like [this](https://love2d.org/wiki/Tutorial:PhysicsDrawing). It would have led us to find this magical :cat2: / :frog: issue from the get-go...

Also, my bad... :sob: that LOVE2D physics learning curve.. :chart_with_downwards_trend: :chart_with_upwards_trend:

**Comment by rhroberts** [comment]
OK, admittedly not good practice on my part, but this PR has grown to encompass quite a few things...

- Environment variable handling
  - There are currently four env vars implemented:
    1) `FMA_DEBUG` - (default: false) set to true to draw physics bodies and a pixel grid; feel free to add other stuff too
    2) `FMA_MUSIC` - (default: true) set to false to turn off the music
    3) `FMA_SCENE` - (default: "TitleScene") jump to a scene when starting FMA
    4) `FMA_PROD` - (default: false) set to true to ignore all other env vars and run game "normally"
- Weird collision saga
  - Two issues here: 
    1) the frog/cat bodies were being added to world even if they weren't drawn, so physicker would run into them. 
    2) physicker was being loaded twice (once in EnterPatientsScene and again in Navigation Scene) so there were two physics bodies, and he would bump into himself. I added a check in physicker:load() to prevent this. 
- I added some more stuff in utils, did some more reorg, and other odds and ends
- A note on the issue above about physicker being blurry and drawing parts of other frames: at some point in my refactoring I moved `love.graphics.setDefaultFilter("nearest", "nearest")` below where the scenes were loaded, so the physicker was being drawn before this was called. Moving up to the top of main fixed that. 

Other thoughts...

- While jumping to specific scenes with `FMA_SCENE` is very handy, it can mess with the game logic a bit (e.g. how to decide where the patients go w/o EnterPatientsScene?). A game is necessarily path-dependent I guess, so skipping that path has some implications lol. Ideally, all of the path-dependent stuff would be tracked in the game state, and perhaps we can have some logical defaults for when we skip to certain scenes.

Anyway, I'm ready to merge this with everyone's OK @CurtisRT @dmarc3 @halfhorst

**Comment by rhroberts** [comment]
> We might want to consider a generic "debug" mode which will draw all existing bodies like [this](https://love2d.org/wiki/Tutorial:PhysicsDrawing). It would have led us to find this magical cat2 / frog issue from the get-go...
> 
> Also, my bad... sob that LOVE2D physics learning curve.. chart_with_downwards_trend chart_with_upwards_trend

I just added this actually 😁 Got the idea from reading through your kiddo fighter!

**Comment by rhroberts** [comment]
Oooh now i see. I added a DEBUG mode but only drew the bodies on a case-by-case basis. Drawing all of the physics bodies that exist in the world is much better, good call

**Comment by dmarc3** [comment]
>         2. physicker was being loaded twice (once in EnterPatientsScene and again in Navigation Scene) so there were
>             two physics bodies, and he would bump into himself. I added a check in physicker:load() to prevent this.

Ooo should we have a "Delete all existing Bodies" method that we call on scene change? I didn't know this was happening but it makes sense. I'll put this on my todo list. I approve the PR again!

**Comment by rhroberts** [comment]
> Ooo should we have a "Delete all existing Bodies" method that we call on scene change?

This is an interesting question... I guess it depends on the two scenes we're changing between?  Changing b/w EnterPatientsScene and NavigationScene we probably want to keep all the bodies. Same for AlchemyScene since that's only a temporary thing and we don't want to keep destroying and creating bodies over and over (assuming nothing in the AlchemyScene needs to be added to the World and collide with stuff; or if it is we just make a new World for that).

What I did with the physicker is add a `self.isLoaded` variable and then wrap everything inside `physicker:load()` in an `if` statement that checks if it's true (default is false, then set to true after first load). This way it can be persistent across scenes without having two bodies or having to destroy and recreate itself. 

I can definitely see a function to remove all bodies from the World being useful in other cases though!

Alright, I'm gonna go ahead and merge 😛
